src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java

```diff
@@ -248,37 +258,95 @@ public void onResume() {
 
     @Override
     public void stop(Throwable cause) throws Exception {
-      PauseAction.endCurrentPause(getContext().get(FlowNode.class));
+      PauseAction.endCurrentPause(getContextClass(FlowNode.class));
       SonarQubeWebHook.get().removeListener(this);
       getContext().onFailure(cause);
     }
 
     @Override
-    public void onTaskCompleted(String taskId, String taskStatus, @Nullable String qgStatus) {
-      if (taskId.equals(step.taskId)) {
+    public void onTaskCompleted(SonarQubeWebHook.Payload payload, String receivedSignature) {
+      if (payload.getTaskId().equals(step.taskId)) {
         try {
-          PauseAction.endCurrentPause(getContext().get(FlowNode.class));
+          PauseAction.endCurrentPause(getContextClass(FlowNode.class));
           SonarQubeWebHook.get().removeListener(this);
-          log("SonarQube task '%s' status is '%s'", step.taskId, taskStatus);
-          switch (taskStatus) {
-            case WsClient.CETask.STATUS_SUCCESS:
-              log("SonarQube task '%s' completed. Quality gate is '%s'", step.taskId, qgStatus);
-              handleQGStatus(qgStatus);
-              break;
-            case WsClient.CETask.STATUS_FAILURE:
-            case WsClient.CETask.STATUS_CANCELED:
-              getContext().onFailure(new IllegalStateException("SonarQube analysis '" + step.getTaskId() + "' failed: " + taskStatus));
-              break;
-            default:
-              throw new IllegalStateException("Unexpected task status: " + taskStatus);
+          if (step.webhookSecretId != null) {
+            if (validateWebhook(receivedSignature, payload.getPayloadAsString())) {
+              // only execute the checkQualityGate if the webhook is found to be valid (getContext().onFailure() does not interrupt execution)
+              checkQualityGate(payload.getTaskStatus(), payload.getQGStatus());
+            }
```

Comment:
Is it ok that we don't fail the pipeline when the secret doesn't match what we expected? (Honest question.)

Reply:
We do fail the pipeline, validateWebhook() will fail the pipeline. But if validateWebhook() fails the pipeline we do not want to proceed with examining of the webhook payload. Making the pipeline fail does not interrupt execution immediately, and I did not want to throw an exception where a simple log line would suffice.